### PR TITLE
Bring-your-own-VPC is used 

### PR DESCRIPTION
**Pull request by tedteng**
**How to categorize this PR?**

/area ops-productivity
/kind bug
/platform gcp

**What this PR does / why we need it**:
fix fetch vpc, network name from infrastructure status instead of joint string

**Which issue(s) this PR fixes**:
Fixes https://github.com/gardener/gardener-extension-provider-gcp/issues/416

**Special notes for your reviewer**:

**Release note**:

```other operator
fix fetch vpc, network name from infrastructure status 
```

### CODE REVIEW

**Comment by tedteng** [comment]
This PR is ready for review

File: pkg/controller/bastion/actuator_reconcile.go [hunk 1]
```diff
@@ -121,6 +130,38 @@ func (a *actuator) Reconcile(ctx context.Context, bastion *extensionsv1alpha1.Ba
 	return a.Client().Status().Patch(ctx, bastion, patch)
 }
 
+func getNetAndSubnet(ctx context.Context, a *actuator, cluster *extensions.Cluster) (vNet, subnet string, err error) {
+	var infrastructureStatus *gcp.InfrastructureStatus
+	worker := &extensionsv1alpha1.Worker{}
+
+	err = a.Client().Get(ctx, client.ObjectKey{Namespace: cluster.ObjectMeta.Name, Name: cluster.Shoot.Name}, worker)
```
Comment (dkistner) [inline pkg/controller/bastion/actuator_reconcile.go:137]: Why are you fetching the `worker` just to get the `InfrastructureStatus`?
Would it not be better to fetch the `Infrastructure` itself and check if it contain the `InfrastructureStatus` in `.status.providerStatus` section?

Comment (tedteng) [reply pkg/controller/bastion/actuator_reconcile.go:137]: I don't how to mock infra status information on my local cluster to verify it and mock integration test either. then I pick up `worker` and will take a look using `Infrastructure`

Comment (tedteng) [reply pkg/controller/bastion/actuator_reconcile.go:137]: had a private discussion with @dkistner there is some difficulty in the APJ region to download images from `eu.gcr.io` at the moment from home. It is extremely slow to download the terraform image when I develop/test the code in my local environment if I need to use the `Infrastructure` object. and integration test is the same situation.  also confirmed with @dkistner "A shoot can only have a machine (if joined or not) if the Worker resource exists."

File: pkg/controller/bastion/actuator_reconcile.go [hunk 2]
```diff
@@ -121,6 +130,38 @@ func (a *actuator) Reconcile(ctx context.Context, bastion *extensionsv1alpha1.Ba
...
+
+	if infrastructureStatus, err = helper.InfrastructureStatusFromRaw(worker.Spec.InfrastructureProviderStatus); err != nil {
+		return "", "", err
+	}
+
+	if infrastructureStatus.Networks.VPC.Name == "" {
+		return "", "", errors.New("virtual network must be not empty for infrastructure provider status")
+	}
+
+	if len(infrastructureStatus.Networks.Subnets) != 1 {
```
Comment (kon-angelo) [inline pkg/controller/bastion/actuator_reconcile.go:154]: This is not a proper assumption here. There can be multiple subnets for different purposes e.g the internal loadbalancing network. You would need to filter for at least the subnet with purpose [nodes](https://github.com/gardener/gardener-extension-provider-gcp/blob/366f966f8084ba63e4abb52eda54cdacd3e14574/pkg/apis/gcp/types_infrastructure.go#L79).

**Review by kon-angelo — CHANGES_REQUESTED** [review]

File: pkg/controller/bastion/actuator_reconcile.go (hunk 2 above)
Comment (tedteng) [reply pkg/controller/bastion/actuator_reconcile.go]: committed in [8194f18](https://github.com/gardener/gardener-extension-provider-gcp/pull/436/commits/8194f18025cd6300c4ca750c0872045595e96801)

File: pkg/controller/bastion/actuator_reconcile.go [hunk 3]
```diff
@@ -121,6 +131,56 @@ func (a *actuator) Reconcile(ctx context.Context, bastion *extensionsv1alpha1.Ba
...
+		return "", "", errors.New("subnet must be not empty")
+	}
+
+	var purposeNodesSet = sets.NewString()
+	for _, subnet := range infrastructureStatus.Networks.Subnets {
+		purposeNodesSet.Insert(string(subnet.Purpose))
+	}
+
+	if !purposeNodesSet.Has(string(gcp.PurposeNodes)) {
+		return "", "", errors.New("none of subenet purpose as node found")
```
Comment (kon-angelo) [inline pkg/controller/bastion/actuator_reconcile.go:165]: Otherwise small nit: `subenet` typo here

Comment (tedteng) [reply pkg/controller/bastion/actuator_reconcile.go]: commit in [1766ec8](https://github.com/gardener/gardener-extension-provider-gcp/pull/436/commits/1766ec88d56f71912d6a0aee977245f7ad01ba46)

[the remark this answers is not in the document]
Comment (tedteng) [reply pkg/controller/bastion/actuator_reconcile.go]: Thanks for the explanation. agree it change in commit https://github.com/gardener/gardener-extension-provider-gcp/pull/436/commits/1766ec88d56f71912d6a0aee977245f7ad01ba46

**Review by kon-angelo — APPROVED** [review]
/lgtm

**Pull request merged**